### PR TITLE
Kademlia fixes

### PR DESCRIPTION
**Pull request by janos**
This PR implements:

- Fixes for Kademlia SuggestPeer and full methods
- Updates Kademlia tests to address the changes
- Adds tests TestKademliaCase* for validating found problems
- Changes NewPeerPot to NewPeerPotMap and removes the need for discover.NodeID slice
- Updates swarm/network/stream/testing snapshots with the ones generated with the updated kademlia SuggestPeer
- Fixes minor lint errors

### CODE REVIEW

**Review by zelig — APPROVED** [review]

File: swarm/network/kademlia.go [hunk 1]
```diff
@@ -677,20 +677,33 @@ func (k *Kademlia) saturation(n int) int {
 func (k *Kademlia) full(emptyBins []int) (full bool) {
```
Comment (nolash) [inline swarm/network/kademlia.go:679]: Please add an explanation of the algorithm used here.

File: swarm/network/kademlia_test.go [hunk 2]
```diff
@@ -405,3 +406,246 @@ func TestKademliaHiveString(t *testing.T) {
...
+	as := make([][]byte, len(addrs))
+	for i, a := range addrs {
+		as[i] = common.FromHex(a)
+	}
+
+	for _, a := range as {
+		if bytes.Equal(a, addr) {
+			continue
+		}
+		p := &BzzAddr{OAddr: a, UAddr: a}
```
Comment (nolash) [inline swarm/network/kademlia_test.go:428]: should oaddr and uaddr be the same?

File: swarm/network/kademlia_test.go [hunk 3]
```diff
@@ -405,3 +406,246 @@ func TestKademliaHiveString(t *testing.T) {
...
+	}
+
+	h := k.Healthy(pp)
+	if !(h.GotNN && h.KnowNN && h.Full) {
+		t.Error("not healthy")
+	}
+}
+
+/*
+The regression test for the following invalid kademlia edge case.
```
Comment (nolash) [inline swarm/network/kademlia_test.go:453]: Why are exactly these cases selected, and where do they come from? Is there any specific logic behind the configurations?

File: swarm/network/stream/snapshot_sync_test.go [hunk 4]
```diff
@@ -246,12 +247,13 @@ func runSyncTest(chunkCount int, nodeCount int, live bool, history bool) error {
...
-				pp := ppmap[id]
+				addr := common.Bytes2Hex(network.ToOverlayAddr(id.Bytes()))
+				pp := ppmap[addr]
 				//call Healthy RPC
 				h := r.delivery.overlay.Healthy(pp)
 				//print info
-				log.Debug(r.delivery.overlay.String())
-				log.Debug(fmt.Sprintf("IS HEALTHY: %t", h.GotNN && h.KnowNN && h.Full))
+				log.Error(r.delivery.overlay.String())
+				log.Error(fmt.Sprintf("IS HEALTHY: %t", h.GotNN && h.KnowNN && h.Full))
```
Comment (nolash) [inline swarm/network/stream/snapshot_sync_test.go]: Should be Debug or Trace?

**Review by nolash — CHANGES_REQUESTED** [review]

File: swarm/network/kademlia_test.go (hunk 2 above)
Comment (janos) [reply swarm/network/kademlia_test.go:428]: They can be for this test.

File: swarm/network/kademlia_test.go (hunk 3 above)
Comment (janos) [reply swarm/network/kademlia_test.go:453]: This cases are the ones discovered as the edge cases while working on other tests.

Comment (nolash) [reply swarm/network/kademlia_test.go:453]: ok, please add a comment saying that.

Comment (janos) [reply swarm/network/kademlia_test.go:453]: I have added the comments. Would more details then this help?

File: swarm/network/kademlia.go (hunk 1 above)
Comment (nolash) [reply swarm/network/kademlia.go:679]: So, as we talked about in the call, please add to the comment the definition of a node being healthy when "as healthy as can be," that is that it can have 0 connected peers if no peers are known in the bin. That's what threw me off, at least.
